Queues: tidy queue call, oracles, data gen, FIFOs

anshumanmohan opened the pull request:
Our queues currently share the same interface, and can therefore we run by the same Calyx program. This Calyx program is generated via `queue_call.py`: it adds a `main` component and makes the queue in question a cell of the main component. That main component also interfaces with the outside world (via .data files) and actually gathers up two memories, representing commands and values. It then invokes the queue in question repeatedly, and stores any answers from the queue component in a third memory. 

This PR tidies up the queue-calling system described above, the data gen script that crafts long lists of commands and values for us, and our Python-level oracles that read those data files and generate trustworthy .expect files. 
- We can now pass important parameters like the number of commands to use and the length of the queue as command-line arguments. 
- We now have an additional flag, `keepgoing`, that tells the oracles or the queue-calling `main` to keep attempting further operations even in the face of overflow or underflow.
- Crucially, we can now turn off a flag during data gen that was creating only error-free sequences of commands. We still keep `no-err` on for SDN, our Piezo work, but otherwise we turn off `no-err` and turn on `keepgoing`.

Finally, I just took the opportunity to also tidy up FIFOs a little.

polybeandip commented on calyx-py/test/correctness/queues/fifo.py:
```diff
@@ -27,75 +27,54 @@ def insert_fifo(prog, name, queue_len_factor=QUEUE_LEN_FACTOR):
+        # Did the user call pop/peek?
+        cb.if_with(fifo.lt_use(cmd, 2), pop_peek_logic),
+        # Did the user call push?
+        cb.if_with(fifo.eq_use(cmd, 2), push_logic),
```
Unless I'm missing something, it seems like we're not raising the error flag when `cmd = 3` here?

anshumanmohan replied:
Yes, that's a lesson I ought to learn from you. Will do!

anshumanmohan replied:
Done in [bcee73e](https://github.com/calyxir/calyx/pull/2163/commits/bcee73ed9982094e3310fa9cd6d187e554dd4314)

polybeandip commented on calyx-py/test/correctness/queues/fifo.py:
```diff
@@ -27,75 +27,54 @@ def insert_fifo(prog, name, queue_len_factor=QUEUE_LEN_FACTOR):
+    # The user called pop/peek.
+    # If the queue is empty, we should raise an error.
+    # Otherwise, we should proceed with the core logic
+    pop_peek_logic = cb.if_with(
```
🤌

polybeandip reviewed (approved):
I think it all looks good to me! I left two minor questions on `fifo.py` and `fifo.data`

polybeandip commented on calyx-py/test/correctness/queues/fifo.data:
The PIFO and FIFO data/expect file were generated with `keepgoing` turned on and with overflow/underflow enabled (i.e. piezo mode turned off), correct?

anshumanmohan replied:
Yes, this PR's secret reason for even existing is to be able to do that! You can see the change being made [here](https://github.com/calyxir/calyx/pull/2163/files#diff-f7b13e9cc848dc4e8df3bab83ea06dec709e95eb9c208b30ac4f2dd97c85dde2)!

anshumanmohan commented:
Thanks for the review. I have resolved your nice catch and will now turn automerge on.

anshumanmohan commented:
Thanks for your review; I addressed your nice catch. I also snuck in a quick additional thing that should have been here all along: the eDSL generators of queues should have been taking a `--keepgoing` flag but were instead hardcoding it to `True`. I have added that in [2b00ce5](https://github.com/calyxir/calyx/pull/2163/commits/2b00ce515929ad5ec5065be33c2a8bc4b7a21413). Going to enable auto-merge!
